builtin/rebase.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * "git rebase" builtin command
+ *
+ * Copyright (c) 2018 Pratik Karki
+ *
+ */
+
+#include "builtin.h"
+#include "run-command.h"
+#include "exec-cmd.h"
+#include "parse-options.h"
+/*
+static const char * const rebase_usage[] = {
+	NULL
+};
+
+static struct option rebase_options[] = {
+	OPT_END()
+};
+*/
+static int use_builtin_rebase(void) {
+	struct child_process cp = CHILD_PROCESS_INIT;
+	struct strbuf out = STRBUF_INIT;
+	int ret;
+
+	argv_array_pushl(&cp.args,
+			 "config", "--bool", "rebase.usebuiltin", NULL);
+	cp.git_cmd = 1;
+	if (capture_command(&cp, &out, 28))
+		return 0;
+
+	strbuf_trim(&out);
+	ret = !strcmp("true", out.buf);
+	strbuf_release(&out);
+	return ret;
+}
+
+int cmd_rebase(int argc, const char **argv, const char *prefix)
+{
+	if (!use_builtin_rebase()) {
+		const char *path = mkpath("%s/git-legacy-rebase",
+					  git_exec_path());
+
+		if (sane_execvp(path, (char **)argv) < 0)
+			die_errno("could not exec %s", path);
+
+		return 0;
+	}
+	prefix = setup_git_directory();
+	trace_repo_setup(prefix);
+	setup_work_tree();
+
+	die("NEEDS WORK");
```

Comment:
That's okay, but I think it would be better to use either `"TODO"` or `"TBD"` (for "To Be Done") here.

Now, let's plan for the next steps because we are really close to the exciting part! I am fairly certain that you can *already* implement enough code to run `git rebase <base>` in a relatively small second commit: As I mentioned a couple of times via mail and IRC (i.e. too far from the code to be useful), what we need to do here is essentially emulate https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/git-legacy-rebase.sh#L203-L222 which kind of boils down to

```sh
. git-rebase--$type
git_rebase__$type${preserve_merges:+__preserve_merges}
```

This is very simple shell scripting that we can call ourselves *very easily* by imitating what `rebase -i`'s `exec` command does: https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/sequencer.c#L2501-L2509

Note: I chose that `exec` example because it not only demonstrates how to execute a Unix shell script snippet, but also how to pass environment variables to it (which you will need to do later, quite a few variables even, because the `git-rebase--<type>` backends have to consume pretty much all of the parsed command-line options in the form of shell variables.

Please do take the time, though, to implement this environment variable passing using `argv_array_push()` and `argv_array_pushf()`, as demonstrated e.g. here: https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/column.c#L377-L394

By the way, it's not quite the truth that we can get away with implementing code to call *just* those two Unix shell commands: there is a function `finish_rebase()` which also needs to be implemented. However, in the first run, it does not need to handle the autostash. And the `rm -f` command corresponds to the POSIX function `unlink()`, the `rm -rf` to `remove_dir_recursively()`, and the most concise example how to run `git gc --auto` is [in `builtin/merge.c`](https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/builtin/merge.c#L408-L416).

To support `git -c rebase.useBuiltin=true rebase <base>`, I think that a very simple "option parsing" would be enough: `if (argc != 2) die("Usage: %s <base>", argv[0]);`.

The autostash part can be handled in a later patch, and should definitely leverage the [`apply_autostash()`](https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/sequencer.c#L2913-L2955) function that was introduced to support finishing interactive rebases in the sequencer; to make that work with the `type=am` mode, you will want to use [`get_dir()`](https://github.com/prertik/git/blob/41531d0ba88cfe20e870d57ee437d0fcd7d05064/sequencer.c#L197-L202) and teach it about the `rebase-apply` directory, and you will want to make `apply_autostash()` non-static (but keep `get_dir()` static).

I really think that we are very close to having something that is both exciting ("It can run a (super-simple) rebase already!") as well as helpful, as it leads the way to using the `t34*.sh` suite of tests to assess how far along this project is. You could, for example, append the line `git config rebase.useBuiltin true` to the body of the `test_create_repo` function in `t/test-lib-functions.sh`, and then run `make -C t t3400-rebase.sh` to see whether any test case succeeds ;-)